guides/source/working_with_javascript_in_rails.md

```diff
@@ -3,561 +3,274 @@
+### Adding NPM Packages with importmap-rails
 
-When clicked, the link background will become red. Here's the problem: what
-happens when we have lots of JavaScript we want to execute on a click?
+To add new packages to your import map-powered application, run the `bin/importmap pin` command
+from your terminal:
 
-```html
-<a href="#" onclick="this.style.backgroundColor='#009900';this.style.color='#FFFFFF';event.preventDefault();">Paint it green</a>
+```bash
+bin/importmap pin react react-dom
 ```
 
-Awkward, right? We could pull the function definition out of the click handler,
-and turn it into a function:
-
-```js
-window.paintIt = function(event, backgroundColor, textColor) {
-  event.preventDefault();
-  event.target.style.backgroundColor = backgroundColor;
-  if (textColor) {
-    event.target.style.color = textColor;
-  }
-}
+Then, import the package into `application.js` as usual:
+```javascript
+import React from "react"
+import ReactDOM from "react-dom"
 ```
 
-And then on our page:
-
-```html
-<a href="#" onclick="paintIt(event, '#990000')">Paint it red</a>
-```
+Adding NPM Packages with JavaScript Bundlers
+--------
 
-That's a little bit better, but what about multiple links that have the same
-effect?
+Import maps are the default for new Rails applications, but if you prefer traditional JavaScript
+bundling, you can create new Rails applications with your choice of
+[esbuild](https://esbuild.github.io/), [webpack](https://webpack.js.org/), or
+[rollup.js](https://rollupjs.org/guide/en/).
 
-```html
-<a href="#" onclick="paintIt(event, '#990000')">Paint it red</a>
-<a href="#" onclick="paintIt(event, '#009900', '#FFFFFF')">Paint it green</a>
-<a href="#" onclick="paintIt(event, '#000099', '#FFFFFF')">Paint it blue</a>
-```
-
-Not very DRY, eh? We can fix this by using events instead. We'll add a `data-*`
-attribute to our link, and then bind a handler to the click event of every link
-that has that attribute:
-
-```js
-function paintIt(element, backgroundColor, textColor) {
-  element.style.backgroundColor = backgroundColor;
-  if (textColor) {
-    element.style.color = textColor;
-  }
-}
-
-window.addEventListener("load", () => {
-  const links = document.querySelectorAll(
-    "a[data-background-color]"
-  );
-  links.forEach((element) => {
-    element.addEventListener("click", (event) => {
-      event.preventDefault();
-
-      const {backgroundColor, textColor} = element.dataset;
-      paintIt(element, backgroundColor, textColor);
-    });
-  });
-});
+To use a bundler instead of import maps in a new Rails application, pass the `—javascript` or `-j`
+option to `bin/rails new`:
 ```
-```html
-<a href="#" data-background-color="#990000">Paint it red</a>
-<a href="#" data-background-color="#009900" data-text-color="#FFFFFF">Paint it green</a>
-<a href="#" data-background-color="#000099" data-text-color="#FFFFFF">Paint it blue</a>
+bin/rails new my_new_app --javascript=webpack
+OR
+bin/rails new my_new_app --javascript=webpack
```

Comment:
```suggestion
bin/rails new my_new_app -j=webpack
```
Sorry, one last thing. I just realized these two lines are the same.

Reply:
Nice catch, thanks!

Also, looking at this again, should these commands just be `rails new` instead of `bin/rails new`? I've always used `bin/rails` for app-level commands like `bin/rails server`, not for `rails new`. Looking at other guides and it seems that almost all use `rails new`

Reply:
You're right. Let's go with `rails new`. My mistake ❤️

Reply:
Perfect, made that change and fixed the `-j` option. Also fixed up bash formatting to lead with a $, matching the rest of the guides.

Thanks for all your help on this!